static/app/utils/useUserTeams.tsx

```diff
@@ -0,0 +1,65 @@
+import {useEffect, useMemo} from 'react';
+
+import OrganizationStore from 'sentry/stores/organizationStore';
+import TeamStore from 'sentry/stores/teamStore';
+import {useLegacyStore} from 'sentry/stores/useLegacyStore';
+import type {Team} from 'sentry/types';
+import {isActiveSuperuser} from 'sentry/utils/isActiveSuperuser';
+import {ApiQueryKey, useApiQuery} from 'sentry/utils/queryClient';
+
+interface UseTeamsResult {
+  isError: boolean | null;
+  isLoading: boolean;
+  teams: Team[];
+}
+
+function buildUserTeamsQueryKey(orgSlug: string): ApiQueryKey {
+  return [`/organizations/${orgSlug}/user-teams/`];
+}
+
+/**
+ * Fetches the user's teams if they aren't already in the TeamStore.
+ * @returns an array of user's teams, isLoading, and isError
+ *
+ * example usage:
+ * ```ts
+ * const {teams, isLoading, isError} = useUserTeams();
+ * ```
+ */
+export function useUserTeams(): UseTeamsResult {
+  const {organization} = useLegacyStore(OrganizationStore);
+  const storeState = useLegacyStore(TeamStore);
+
+  // Wait until the store has loaded to start fetching
+  const shouldConsiderFetching = !!organization?.slug && !storeState.loading;
+  // Only fetch if there are missing teams
+  const hasMissing = !storeState.loadedUserTeams;
+  const queryEnabled = shouldConsiderFetching && hasMissing;
+
+  const {
+    data: additionalTeams = [],
+    isLoading,
+    isError,
+  } = useApiQuery<Team[]>(buildUserTeamsQueryKey(organization?.slug ?? ''), {
```

Comment:
I wonder if it's even worth using useApiQuery here since we aren't utilizing many of the benefits. Not using the cache, since the response is put in TeamStore. Not utilizing the refetch in background since we only want this to run when the TeamStore says it needs something.

I think you've configured it correctly, so no need to change anything, just food for thought on how we want to build these kind of hooks with special caching logic in the future (with or without react query)

Reply:
Although, I guess this does still use the request de-duping feature which may be why this new hook no longer makes duplicate requests. So I suppose that is one big reason to use it over a basic fetch

Reply:
this is mostly for de-duping